[WIP] Implement cpu pinning and cgroup cpuset support

poelzi opened the pull request:
Add command line argument to specify a cpuid and cpuset
cgroup name for the engine.
The engine tries at startup to move itself to the cpuset
provided or set the cpu affinity.
This will allow a wrapper program to free a core for the engine,
minimizing latency jitter or CPU starvating.

## Instructions

after installing https://github.com/lpechacek/cpuset/releases/tag/v1.6
start with mixxx-isolate
`tools/mixxx-isolate.py -- [mixxx args]`

The script will start itself as root in the background to do the necessary system changes. After that, it starts
mixxx which will itself move the engine into a special cpuset group.
After mixxx exists, the system is restored.
The script also takes the hyperthreading cpus offline that are siblings of the engine cpu.
This way, you get the most power out of your CPU but have none of the negative effects of hyperthreading.
You can customize the behaviour:

```
usage: mixxx-isolate.py [-h] [--cpuset_engine CPUSET_ENGINE] [--cpuset_system CPUSET_SYSTEM] [--cpuset_isolation CPUSET_ISOLATION] [--no-isolate-ht] [--no-offline] [--cpu-id CPU] [--cleanup] [--mixxx MIXXX] [--debug]
                        [mixxx_args [mixxx_args ...]]

Runs mixxx with isolated engine on a dedicated CPU. You can pass mixxx arguments using ' -- 'after mixxx-isolate arguments. This program will seperate one CPU core from the system and place the mixxx engine thread on a dedicated CPU.
Hyperthreading siblings will be isolated and taken offline if possible. The normal system state is restored when mixxx is closed.

positional arguments:
  mixxx_args            Arguments passed to mixxx

optional arguments:
  -h, --help            show this help message and exit
  --cpuset_engine CPUSET_ENGINE
                        cpuset name for the mixxx engine
  --cpuset_system CPUSET_SYSTEM
                        cpuset name for the mixxx engine
  --cpuset_isolation CPUSET_ISOLATION
                        cpuset name for the mixxx engine
  --no-isolate-ht       don't isolate hyperthreading partner (not recommended)
  --no-offline          don't take hyperthreading cpus offline
  --cpu-id CPU          ID of the cpu to use for the engine. -1 for autoselect
  --cleanup             Cleanup old isolation settings
  --mixxx MIXXX         Mixxx executable
  --debug               Debug output
```

## Manual Instructions


You need to find out which cpu id's are one core by looking at /proc/cpuinfo and compare the core id.
In my case I use cpu 3 and it's hyperthreading partner 7

```sh
sudo cset shield --cpu 3,7 --kthread=on
echo 3 | sudo tee -a /sys/fs/cgroup/cpuset/user/cpuset.cpus
sudo chown $(id -u) /sys/fs/cgroup/cpuset/user/tasks
mixxx  --settingsPath ~/Projects/mixxx-devsettings/ --logLevel debug --developer --controllerDebug --engineCpuSet user
```

This logic will be put into mixxx-isolate (python script) and it will try to request the proper credentials through modern desktop techniques. This is only the mixxx part yet.

## Restore system
Stop the shield after you finished your usage:
```
sudo cset shield -r
```

### Find out which CPU to use

Look at the cpuinfo file
```sh
cat /proc/cpuinfo
```
The output looks like this:
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 158
model name      : Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
stepping        : 9
microcode       : 0xde
cpu MHz         : 2839.943
cache size      : 8192 KB
physical id     : 0
siblings        : 8
core id         : 0
cpu cores       : 4
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa itlb_multihit srbds
bogomips        : 5799.77
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:

[...]
```
You can now choose any cpu, and find the matching entries which have the same *core id*. Use those in the instructions above.
Be aware that CPU Ids start at 0.

poelzi committed:
Implement cpu pinning and cgroup cpuset support

Add command line argument to specify a cpuid and cpuset
cgroup name for the engine.
The engine tries at startup to move itself to the cpuset
provided or set the cpu affinity.
This will allow a wrapper program to free a core for the engine,
minimizing latency jitter or CPU starvating.

poelzi commented:
## Instructions

You need to find out which cpu id's are one core by looking at /proc/cpuinfo and compare the core id.
In my case I use cpu 3 and it's hyperthreading partner 7

```sh
sudo cset shield --cpu 3,7 --kthread=on
echo 3 | sudo tee -a /sys/fs/cgroup/cpuset/user/cpuset.cpus
sudo chown $(id -u) /sys/fs/cgroup/cpuset/user/tasks
mixxx  --settingsPath ~/Projects/mixxx-devsettings/ --logLevel debug --developer --controllerDebug --engineCpuSet user
```

This logic will be put into mixxx-isolate (python script) and it will try to request the proper credentials through modern desktop techniques. This is only the mixxx part yet.

geraldog commented:
Hi @poelzi thanks for working on this! Unfortunately as I don't use Ubuntu I don't have "cset" on my system. I searched Google and it seems it went from Google code to https://github.com/lpechacek/cpuset

Do you think that's the right reference and it's safe to download code from that repository?

poelzi commented:
@geraldog yes, thats the repo. mixxx-isolate will not use this script later but does the work itself. For testing this script makes the necessary work.

daschuer commented on src/engine/enginemaster.cpp:
```diff
@@ -381,11 +382,25 @@ void EngineMaster::processChannels(int iBufferSize) {
+void EngineMaster::finishStartup() {
+    QThread::currentThread()->setObjectName("Engine");
+#ifdef __LINUX__
+    const auto cpuSet = CmdlineArgs::Instance().getEngineCpuSet();
```
Even though this is only a one time issue, I would prefer not to use Instance() from the engine thread. 
Can we cache the result in a QAtomicInt and use it here?

daschuer replied:
Wait, what is auto? Please don't use auto here. The type is out of sight.

daschuer commented on src/util/cmdlineargs.cpp:
```diff
@@ -178,6 +194,9 @@ void CmdlineArgs::printUsage() {
                         %s\n", getSettingsPath().toLocal8Bit().constData());
     fputs("\
 \n\
+--engineCpuId CPUID     Bind mixxx to a specific CPU Core. Use mixxx-isolate instead.\n\
```
```suggestion
--engineCpuId CPUID    Used by mixxx-isolate to bind mixxx to a specific CPU Core.\n\
```

daschuer commented on src/util/cpupinning.cpp:
```diff
@@ -0,0 +1,80 @@
+#ifdef __LINUX__
+bool CpuPinning::moveThreadToCpuset(const QString& cgroup) {
+    qDebug() << "Move task" << gettid() << "to cpuset" << cgroup;
+    QString filename = QLatin1String("/sys/fs/cgroup/cpuset/") + cgroup + QLatin1String("/tasks");
```
Use QStringLiteral when creating strings, this avoids utf8 to utf16 conversion at runtime.

Holzhaus replied:
Yes, QLatin1String should only be used when you don't actually construct a qstring, e.g. when using comparisons like `str.startsWith(QLatin1String("foo"))`.

daschuer commented on src/util/cpupinning.cpp:
```diff
@@ -0,0 +1,80 @@
+    pthread_t thread = pthread_self();
+    cpu_set_t cpuset;
+    int status;
+    /* Set affinity mask to include CPUs 0 to 7 */
```
Use cpp comments.  0 to n

daschuer commented on src/util/cpupinning.cpp:
```diff
@@ -0,0 +1,80 @@
+#ifdef __LINUX__
+    qDebug() << "Set CPU affinity for Thread ID" << gettid();
+
+    pthread_t thread = pthread_self();
```
I think it would be better do do this from outside the engine thread.

daschuer commented on src/util/cpupinning.cpp:
```diff
@@ -0,0 +1,80 @@
+
+#ifdef __LINUX__
+bool CpuPinning::moveThreadToCpuset(const QString& cgroup) {
+    qDebug() << "Move task" << gettid() << "to cpuset" << cgroup;
```
This dos not build. 
/home/daniel/workspace/advanced_autodj_2/src/util/cpupinning.cpp:21:32: error: ‘gettid’ was not declared in this scope
     qDebug() << "Move task" << gettid() << "to cpuset" << cgroup;

daschuer replied:
including this works: 
```
#if __GLIBC__ == 2 && __GLIBC_MINOR__ < 30
#include <sys/syscall.h>
#define gettid() syscall(SYS_gettid)
#endif
```
found here: 
https://stackoverflow.com/questions/21279649/getting-error-in-c-program-undefined-reference-to-gettid

daschuer commented:
If you reconfigure the sound hardware the thread id of the audio thread changes.
In this case we need to redo the config.

daschuer commented:
Works good even on my: 
Intel(R) Core(TM) i5-3317U CPU @ 1.70GHz with only two cores + HT. 

It is now immune against CPU loads from other applications. 

Since I tried this last time I had a bad experience due to the missing CPUs, I cannot confirm it this time. 

The scaling of the CPU latency indicator of Mixxx is now not longer significant. I can add a deck with key-lock even if it is at max without overflow. The system monitor only reports ~30 % load. 
The maximum is defined here as 25 %: 
https://github.com/mixxxdj/mixxx/blob/95a5716a54b6bc0f937ea08755bdfc4cd01ee351/src/engine/enginemaster.cpp#L77

geraldog commented:
In my case I have an Intel Core i5-3337U @ 1.80GHz CPU which I believe has four logical cores and hence 2 physical cores. I ran as root:

`./cset shield --cpu 0,1 --kthread=o`
(required Python future)
`echo 0 | tee -a /sys/fs/cgroup/cpuset/user/cpuset.cpus`
`chown 1000 /sys/fs/cgroup/cpuset/user/tasks`

I then ran `mkdir SET; build/mixxx  --settingsPath ./SET --logLevel debug --developer --controllerDebug --engineCpuSet user` which did work and apparently confined the engine thread to CPU 0, at least to my knowledge.

Rubberband on 4 decks still doesn't work here :)

It doesn't work on bloody 2 decks to tell you the truth, the processor is Q1 2013, so 8 years old. Running 5.9.1 - RT gentoo

geraldog commented:
I confirm the issue with Soundtouch is related to the cpuset pinning.

Will start consuming loads of CPU with very distorted sound and this:
```
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
warning [Engine] underflowHappened code: 24
```

Lots of this basically :(

That's again on a 5.9.1 RT kernel, on gentoo.

daschuer commented:
In my test I use cpu 2 and 3.
Does it make a difference?

daschuer commented:
Is it possible to record the distorted sound from a short 440 Hz sine wave?
Maybe looking at it with audacity gives some insights.

geraldog commented:
Hey @daschuer maybe I did something wrong? As I understood from /proc/cpuinfo 0 and 1 are hyperthreading partners for my processor, with core_id = 0

And it works except for Soundtouch. Can somebody else reproduce the bug? I certainly can record the sound if you wish @daschuer

poelzi commented:
@geraldog It depends on the processor which cores are from the same core. Some processors have the HT core next to the main one, some have the normal first and then all the HT cores (like mine). All cpu id's that share a "core id" belong to each other and can cause latency spikes on each other.

poelzi commented:
you can put your cpuinfo into a gist and post it here and I will take a look.

geraldog commented:
Thanks @poelzi here it is https://gist.github.com/geraldog/ee8f6bb36886d69b17396ec2414e3b4d by the way the fact that you and @daschuer are both named Daniel confuses me a lot!

daschuer commented:
Ok 0+1 an 2+3 are partners. This is the same on my machine.

When I use cset it reports that it is not able to live all threads. Maybe something is special with the very first CPU.

Can you reproduce the issue using cpu 2+3?

When I tried this a while ago I had similar experience with ad 2 + 2 cpu, cause you literally reduce your CPU power by 50 %. This ratio becomes better when utilizing a 4 + 4 cpu.
This was on a different CPU than today. Maybe something has changed since then.

Make also sure to noch Tuch the hardware preferences. Applying a chance destroyes the trick, because the engine threads is then moved to the already overcrowded Cpu pair.

Can you confirm that with a system monitor?
If everything works you should see in you original case CPU 1+2 idling around and a higher Cpu load for one of them when enabling RubberBand key lock.

daschuer commented:
If you still have issue please provide a 440 Hz sine wave recording.

geraldog commented:
@daschuer I tried cset with CPUs 2 and 3 and still the same issue with Soundtouch. I recorded from Mixxx the 440hz tone you asked but it didn't record the glitches? I'll have to record with an external recorder.

daschuer commented:
> I'll have to record with an external recorder.

Great, that means that there is nothing inside Mixxx that produces the glitches.

Which sound card and sound API do you use? Maybe there is a second thread involved that is now running out of CPU.

poelzi commented:
Meybe we should move the kernel thread of the soundcard together with jackd if used to this processor as well.

poelzi commented:
@geraldog does the interface tab in the preferences report any buffer underruns ?

poelzi commented:
This might be interesting as well: https://www.alsa-project.org/wiki/Low_latency_howto

geraldog commented:
I take back my previous comment, there's something wrong with Soundtouch on main, CPU pinning only exacerbates it. There are CPU spikes into the red zone and little glitches, which becomes full red when CPU pinning and full glitching.

@poelzi I believe underflows are reported in the Sound Hardware tab, I see nothing under Interface.

geraldog commented:
@daschuer it doesn't matter if I use ALSA or JACK wrapping up ALSA, glitches the same, see previous comment.

geraldog commented:
By the way it also doesn't matter if I use the internal soundcard of the Laptop or my controller's USB soundcard. Glitches the same.

poelzi committed:
Add mixxx-isolate helper script.

Add script which will isolate a dedicated CPU for the mixxx engine.

- Select a CPU and its hyperthread partners (last real cpu)
- Create cpuset groups for the system with all other cpus and move all
processes there.
- create a cpuset group for the engine and one for all hyperthreading partners
- try to take hyperthread cpu offline
- start mixx with provided paramenters and cgroup parameters

when mixxx is stopped
- take cpu online again
- move all processes back to the root group
- delete cpuset groups

poelzi commented:
I wrote a script that does not all the heavy lifting. It is under tools/mixxx-isolate.py 
Please test :)
It now even tries to take the hyperthread cpu offline so there are not even unbound kernel threads running on it :)

poelzi commented:
It requires: https://github.com/lpechacek/cpuset/releases/tag/v1.6 and python3

geraldog commented:
It requires sudo ;-)

geraldog commented:
I did not want to install cpuset to my whole machine so I just copied mixxx-isolate.py to the cpuset-master directory, and ran from there.

`./mixxx-isolate.py --mixxx /home/dumbo/mixxx-cpu-pin/build/mixxx -- --settingsPath /home/dumbo/mixxx-cpu-pin/SET --logLevel debug --developer --controllerDebug`

Works perfectly, thank you @poelzi but the issue with Soundtouch remains for me, and Rubberband unfortunately hits 35% of the isolated CPU on barely two decks :-(

poelzi commented:
I will look if we can also generate a policykit spec that allows you to run the script without password request. If you run it outside a shell, he tries policykit and qsudo for getting root.

Abdull commented:
I tried to build this PR, but CMake aborts because of an incorrect hash for `libKeyFinder/v2.2.3.zip`:

```
$ cmake --build ./build --parallel `nproc`
# ...
-- [download 100% complete]
-- verifying file...
       file='/home/pi/git/mixxx-2021/build/download/libKeyFinder/v2.2.3.zip'
-- SHA256 hash of
    /home/pi/git/mixxx-2021/build/download/libKeyFinder/v2.2.3.zip
  does not match expected value
    expected: '56887e7b51834223d4264f4ea9acf16f0b6a0c706c218478afbaf89ad1f9c6ea'
      actual: 'ad43ca006e3bbed0810ff62e170d04522a64f8606c2166bfa5a9b9158b7ebc11'
-- Hash mismatch, removing...
CMake Error at libKeyFinder-stamp/download-libKeyFinder.cmake:159 (message):
  Each download failed!

make[2]: *** [CMakeFiles/libKeyFinder.dir/build.make:90: libKeyFinder-prefix/src/libKeyFinder-stamp/libKeyFinder-download] Error 1
make[1]: *** [CMakeFiles/Makefile2:361: CMakeFiles/libKeyFinder.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```

In case it's relevant for this problem: I've built another PR before *this* PR within `/home/pi/git/mixxx-2021/` - could this cause the hash mismatch? Checking out the other PR again and running `cmake` does not cause the hash mismatch.

Be-ing commented:
@Abdull merge main to this branch locally.

Abdull commented:
> @Abdull merge main to this branch locally.

Thanks, with that I could make progress, but now I get a scoping issue during build:

```
$ git log -n 1
commit 61d49fa0c30afdaf774897565fe375a5d2769c4a (HEAD -> cpuset)
Author: Daniel Poelzleithner <git@poelzi.org>
Date:   Thu Jan 14 02:38:29 2021 +0100

# my git repository is currently on *this* PR


$ git log -n 1 origin/main
commit 75b52c2bbbadc14a7da57f686091cb708e5e7543 (origin/main, origin/HEAD, main)
Merge: a6baf3f431 c0e03f3bb0
Author: Daniel Schürmann <daschuer@mixxx.org>
Date:   Thu Jan 21 15:22:45 2021 +0100

# the main branch I'm about to merge
# ... now merging that main into this PR branch

$ git merge --no-ff --no-commit origin/main

# ...
Automatic merge went well; stopped before committing as requested

# so far so good, let's prepare:

$ cmake -DCMAKE_INSTALL_PREFIX=/usr/local -S . -B ./build
-- Optimization level: portable
-- Git branch: cpuset
-- Git commit: 61d49fa0c3
-- Git commit count: 7722
-- Found ccache: /usr/bin/ccache
-- Support for ccache: ON
-- git Version: v0.0.0-dirty
-- Version: 0.0.0
-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK -- success
-- Linking libebur128 dynamically
-- Could NOT find KeyFinder (missing: KeyFinder_LIBRARY KeyFinder_INCLUDE_DIR) 
-- Linking libSoundTouch dynamically
-- Could NOT find MP4 (missing: MP4_LIBRARY MP4_INCLUDE_DIR) 
-- Could NOT find MP4v2 (missing: MP4v2_LIBRARY MP4v2_INCLUDE_DIR) 
-- Could NOT find HSS1394 (missing: HSS1394_LIBRARY HSS1394_INCLUDE_DIR) 
-- Installed libshout version: 2.4.1 is suffering from bug lp1833225
-- Using internal libshout
-- Linking libhidapi dynamically
-- Configuring done
-- Generating done
-- Build files have been written to: /home/pi/git/mixxx-2021/build

# ...

$ cmake --build ./build --parallel `nproc`
# ...
/home/pi/git/mixxx-2021/src/util/cpupinning.cpp: In static member function ‘static bool mixxx::CpuPinning::moveThreadToCpuset(const QString&)’:
/home/pi/git/mixxx-2021/src/util/cpupinning.cpp:21:32: error: ‘gettid’ was not declared in this scope
     qDebug() << "Move task" << gettid() << "to cpuset" << cgroup;
                                ^~~~~~
/home/pi/git/mixxx-2021/src/util/cpupinning.cpp:21:32: note: suggested alternative: ‘getgid’
     qDebug() << "Move task" << gettid() << "to cpuset" << cgroup;
                                ^~~~~~
                                getgid
/home/pi/git/mixxx-2021/src/util/cpupinning.cpp: In static member function ‘static bool mixxx::CpuPinning::pinThreadToCpu(qint32)’:
/home/pi/git/mixxx-2021/src/util/cpupinning.cpp:41:53: error: ‘gettid’ was not declared in this scope
     qDebug() << "Set CPU affinity for Thread ID" << gettid();
                                                     ^~~~~~
/home/pi/git/mixxx-2021/src/util/cpupinning.cpp:41:53: note: suggested alternative: ‘getgid’
     qDebug() << "Set CPU affinity for Thread ID" << gettid();
                                                     ^~~~~~
                                                     getgid
make[2]: *** [CMakeFiles/mixxx-lib.dir/build.make:5224: CMakeFiles/mixxx-lib.dir/src/util/cpupinning.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1183: CMakeFiles/mixxx-lib.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```

poelzi committed:
fix ubuntu build

ronso0 commented:
@poelzi Do you still maintain this branch?

I wanted to do some testing on top of `main` so I resolved the conflicts and fixed the script.
for anyone who want's to test it, https://github.com/ronso0/mixxx/tree/cpu-pinning is my updated branch

nothing more than
f941f2437c merge main and resolved the conflicts 
d4a5375686 fix the script to use 'engine-cpu-set'

Will come back after some real-life testing :)

ronso0 commented:
> I did not want to install cpuset to my whole machine so I just copied mixxx-isolate.py to the cpuset-master directory, and ran from there.
> 
> `./mixxx-isolate.py --mixxx /home/dumbo/mixxx-cpu-pin/build/mixxx -- --settingsPath /home/dumbo/mixxx-cpu-pin/SET --logLevel debug --developer --controllerDebug`

@geraldog Thanks for the hint!

ronso0 commented:
Also, for the wishlist:
make `mixxx-isolate.py` available when running mixxx from the build directiory while _not_ having checked out this branch in the source directory.
Simply copy `mixxx-isolate.py` to the build dir when building?

ronso0 commented:
Some feedback from recent real-life testing:
when running Mixxx I sometimes set the CPU scaling governor to 'performance', sometimes I forget and it remains 'ondemand'.
When I forget and use CPU pinning the performance with keylock much worse and I need to double the buffer to avoid xruns.

With 'performance' I didn't notice a huge difference, I could not lower the buffer without xruns.
I'll keep testing..

ronso0 commented:
@poelzi 
Do you still intent to bring this into a mergable state?
If no, I'll fix the conflicts and push to this branch so it can be tested more easily.

poelzi commented:
Yes, I actually started to rewrite the cgroups helper in rust and also do support cgroups2 properly.
I also want it to be able to do it after mixxx has started. When I come to the point of implementing live mode, it should activate automatically.

poelzi commented:
Yes, I actually started to rewrite the cgroups helper in rust and also do support cgroups2 properly.
I also want it to be able to do it after mixxx has started. When I come to the point of implementing live mode, it should activate automatically.

ronso0 commented:
> Some feedback from recent real-life testing:
when running Mixxx I sometimes set the CPU scaling governor to 'performance', sometimes I forget and it remains 'ondemand'.
When I forget and use CPU pinning the performance with keylock much worse and I need to double the buffer to avoid xruns.
>
> With 'performance' I didn't notice a huge difference, I could not lower the buffer without xruns.
I'll keep testing..

With this merged to `main` and all CPU governors set to `performance` I can run 3 decks with effects and pitch != 0 (RubberBand) with 5ms stable :tada:

ronso0 commented:
FYI I ported this to `main` #12806

ronso0 commented:
Superseded by #12806
